File: Sorting and Searching/Josephus Problem II/JosephusProblemII.cpp
// CSES Problem Set - Sorting and Searching - Josephus Problem II (https://cses.fi/problemset/task/2163/)
// Solution by: nagajas

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp> 
#include <ext/pb_ds/tree_policy.hpp>

using namespace __gnu_pbds;
using namespace std;

#define ordered_set tree<int, null_type,less<int>, rb_tree_tag,tree_order_statistics_node_update> 

int main(){
    int n,k;
    scanf("%d %d",&n,&k);
    
    ordered_set s;
    for(int i=1;i<=n;i++) s.insert(i);
    int prev = 0;
    for(int i=0;i<n-1;i++){
        int curr = (prev+k)%(n-i);
        prev = curr;
        
        auto y = s.find_by_order(curr);
        printf("%d ",*y);
        s.erase(y);
    }
    printf("%d\n",*s.begin());
    //printf("time taken : %f secs\n", (float)clock() / CLOCKS_PER_SEC);
    return 0;
}
// Result: ACCEPTED

File: Sorting and Searching/Josephus Problem II/JosephusProblemII.py
# CSES Problem Set - Sorting and Searching - Josephus Problem II (https://cses.fi/problemset/task/2163/)
# Solution by: nagajas
# Note: This is not an accepted solution, please see the footer for more details.

from sortedcontainers import SortedSet

n, k = map(int, input().split())
soldiers = SortedSet(range(1, n + 1))
prev = 0
for i in range(n-1):
    curr = (prev + k) % (n-i)
    prev = curr
    print(soldiers[curr], end=" ")
    soldiers.remove(soldiers[curr])
print(soldiers[0])

## This version is simplified and readable version of accepted solution.
## This is not accepted because of the use of sortedcontainers module, which is not available in the judge.
## Accepted solution uses complete implementation of SortedSet (from the module) without importing any module.
## Refer to ./JosephusProblemII_full.py for the accepted solution.